### PR TITLE
Update the configure script for OCaml 5.00

### PR DESCRIPTION
**Pull request by xavierleroy**
This PR is a collection of updates to configure.ac that reflect the current status of OCaml 5.00.

See the individual commits for a detailed description.

Feel free to suggest other things that need  updating in configure.

Cc: @shindere @dra27

### CODE REVIEW

**Comment by shindere** [comment]
Xavier Leroy (2022/01/11 01:46 -0800):
> This PR is a collection of updates to configure.ac that reflect the current status of OCaml 5.00.
>
> See the individual commits for a detailed description.
>
> Feel free to suggest other things that need  updating in configure.
>
> Cc: @shindere @dra27
>
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/ocaml/ocaml/pull/10871
>
> -- Commit Summary --
>
>   * configure: update the --disable-systhreads option
>   * configure: reject --enable-flambda
>   * configure: more careful configuration for the native-code compiler
>   * configure: update the list of configurations that support native code
>
> -- File Changes --
>
>     M configure (267)
>     M configure.ac (235)
>     M tools/ci/inria/other-configs/script (1)
>
> -- Patch Links --
>
> https://github.com/ocaml/ocaml/pull/10871.patch
> https://github.com/ocaml/ocaml/pull/10871.diff
>
ery nice work, many thanks!

About the message

```
POSIX threads are required but not supported on this platform
```

I wonder whether we should say somehting like `POSIX threads are
necessary to build the runtime`, to make it clear that it's not that
they have been requested by the user or so. What do you think?

**Comment by shindere** [comment]
About the flambda change:

- Shouldn't the `--enable-flambda-invariants` be treated the same way?

At the moment the configure script accepts it.

You may want to disable the flambda check in Inria's other-configs job,
that is, in `tools/ci/inria/other-configs/script`.

Lastly, do you think it's worth docuemnting in the helpstirng of the
configure options that they are not supported?

**Comment by lthls** [comment]
Is there an issue somewhere tracking the status of flambda support ? I'm a bit disappointed that we seem to be considering a release without flambda support, and if the only issues are mysterious error messages I could probably find some time to investigate before the release.

**Comment by gasche** [comment]
> I'm a bit disappointed that we seem to be considering a release without flambda support

I would hazard the guess that this is mostly a matter of contributors/maintainers with flambda knowledge chiming in and looking at what needs to be fixed, if any. There is ample time left before a release, but we need people volunteering to do the work.

My (non-expert) understanding is the same as yours: some new code plays badly with flambda's mutability inference (which is already known to not be fully compatible with GADTs for example), and it's a matter of massaging it, or finding out how to tune the flambda analysis. But there should be nothing fundamentally incompatible between multicore and flambda.

**Comment by gasche** [comment]
I created an issue to track Flambda support in trunk: #10878.

@xavierleroy unless I'm missing something, I see an issue with the present PR for people who wish to contribute better support for flambda, or for currently not-supported architectures. How are they supposed to configure the compiler to experiment with the currently-broken support? If `--enable-flambda` is gone, how should I try to build with Flambda to fix the issues?

(Naturally I would expect the options to still exist and put me in a half-broken state, printing a warning that the configuration is currently not support for users. Or maybe it could fail for release builds, and just print the warning for development builds?)

(I guess that the PR's secret motive was to motivate people to work on better support; well done :-)

**Comment by shindere** [comment]
No, I think the motivation for the PR was to make CI skip those
architectures or configurations where OCaml is not supported yet.

**Comment by xavierleroy** [comment]
The purpose  of the present PR is to avoid reports such as #10869 from unsuspecting users who try to build the trunk with unsupported configurations and end up with mysterious errors that they report as bugs.  If something is known to be broken, and  is planned to stay broken for the next release, it should not be configurable.  

> I'm a bit disappointed that we seem to be considering a release without flambda support

I think this was the plan for the "minimal viable product", i.e. release 5.00.  But if flambda is made to work before the release, it's even better.

> How are they supposed to configure the compiler to experiment with the currently-broken support? If --enable-flambda is gone, how should I try to build with Flambda to fix the issues?

After configure, edit Makefile.config by hand or by script, replacing `FLAMBDA=false` by `FLAMBDA=true`.

Or: uncomment the flambda-related code in configure.ac (the one that this PR comments out), make configure, configure --enable-flambda.

**Comment by xavierleroy** [comment]
> Shouldn't the `--enable-flambda-invariants` be treated the same way?

The old configure script accepts and ignores `--enable-flambda-invariants` even if flambda isn't activated.  I didn't change that.

> Lastly, do you think it's worth docuemnting in the helpstirng of the configure options that they are not supported?

We hope these options to be back in 5.01 or 5.02.  Again I'm shooting for minimal changes.

**Comment by shindere** [comment]
Xavier Leroy (2022/01/11 09:11 -0800):
> After configure, edit Makefile.config by hand or by script, replacing `FLAMBDA=false` by `FLAMBDA=true`.
>
> Or: uncomment the flambda-related code in configure.ac (the one that
> this PR comments out), make configure, configure --enable-flambda.

As suggested privately, perhaps we could have a distinction between
developent and release builds.

In developent builds we can let people configure what they like so that
they can work on restauring the new architectures of flambda or whatever
and in release mode yes, we can refuse to configure on those platforms
that are known not to be supported.

**Comment by xavierleroy** [comment]
> As suggested privately, perhaps we could have a distinction between development and release builds.

Users also build from trunk, see #10869 for an example.

**Comment by shindere** [comment]
Xavier Leroy (2022/01/11 09:37 -0800):
> > As suggested privately, perhaps we could have a distinction between development and release builds.
>
> Users also build from trunk, see #10869 for an example.

I would perhaps not be too hard to check an environmnet variable to see
whether unsafe configs can be allowed, like
`OCAML_ALLOW_UNSAFE_CONFIGS`, or something like that.

**Comment by gasche** [comment]
If users build from the development version, right after a major change that was largely advertised, they shouldn't expect a polished experience. When the code works, we release :-) I'm not sure it's a good deal to make life harder for people wishing to hack on the compiler, right when we would benefit from having people look at what's broken and consider helping fix it.

**Comment by shindere** [comment]
Gabriel Scherer (2022/01/11 12:23 -0800):
> If users build from the development version, right after a major
> change that was largely advertised, they shouldn't expect a polished
> experience. When the code works, we release :-) I'm not sure it's a
> good deal to make life harder for people wishing to hack on the
> compiler, right when we would benefit from having people look at
> what's broken and consider helping fix it.

Yeah, I tend to agree. It's easy to have CI ignore some configurations
without having to change anything to the code outside of the CI scripts.

**Comment by xavierleroy** [comment]
I reverted the removal of the --with-flambda option now that Flambda is back.  However, I still think the 3 remaining commits in this PR are useful, and necessary to un-break Inria's CI.

**Comment by gasche** [comment]
Sorry for the naive question, but out of curiosity: why do we need to change the configure script to un-break the CI? I would naively think that the CI script has several configuration that it tests, and that we could disable/comment the invocations that are currently not working, or explicitly change their configuration to bytecode-only for example.

(I hadn't realized at all that this PR was related to the CI before it was mentioned by @shindere earlier.)

**Comment by xavierleroy** [comment]
> Sorry for the naive question, but out of curiosity: why do we need to change the configure script to un-break the CI?

Because a number of configurations previously tested now break for silly reasons.  E.g. one of them configures with `--disable-systhreads` and it prevents the whole system from being compiled (instead of just the otherlibs/systhreads directory).

File: configure.ac
Comment (gasche) [inline configure.ac]: I think it would be nice to keep these comments positioned in such a way that, if I want to hack on a backend for one of those systems, I can just uncomment the two lines and start running.

File: configure.ac
Comment (gasche) [inline configure.ac:1838]: If I understand correctly, previously we were trying to guess whether systhreads can be enabled (by checking whether (1) Windows or (2) both pthreads and `sigwait` are available), and currently you ask for an explicit setting without auto-detection logic? (Or do you know assume `yes` unless `unix` is known to not be supported?)

(It's a bit sad to see the auto-detection logic go away, but I guess the people that will work on Windows support or whatever will know to just look at the 4.14 configure script to take inspiration from.)

**Review by gasche — APPROVED** [review]
I don't really understand, but I'm glad to have confirmation that I'm too dumb to work on the CI and leave it to other people. In the meantime, here is an approval that the configure.ac changes look, ahem, not actively terrible.

**Comment by shindere** [comment]
If it's just a question of CI, at least as far as Inria's CI is
concerned, I don't think it's true that a modification of the configure
script is required.

I am not saying it is not useful for anything, though, it may have other
uses. But as far as CI is concerned no modification of the configure
script is requred in my opinion and I will submit code soon to
workaroudn the Incria CI issues.

File: configure.ac
Comment (xavierleroy) [reply configure.ac:1838]: POSIX threads or the Win32 equivalent are now required by the runtime system (how many times will I have to write this?), so systhreads can always be built, and it's the default.  It is not built only if the user configured with `--disable-systhreads`.

File: configure.ac
Comment (xavierleroy) [reply configure.ac]: These are shell comments, not m4 comments, so there's a risk of messing with the m4 macros.

File: configure.ac
Comment (dra27) [inline configure.ac]: To please check-typo (the generated `configure` file is unaffected by this change):
```suggestion
    [AC_MSG_ERROR(m4_normalize([POSIX threads are required but not supported on
      this platform]))])])
```

File: configure.ac
Comment (dra27) [reply configure.ac]: I agree that it would be nice to keep the diff quieter - the "appropriate" syntax in this case is to use `dnl` which is m4's comment indicator. I double-checked this in https://github.com/dra27/ocaml/commit/43020f7ab9d4fe69051c79bbed23f0455601b561. By putting `mingw-w64` last, it does mean that deleting the `dnl MVP:` is sufficient to re-add any case, as @gasche requests.

**Review by dra27 — COMMENTED** [review]
Completely agree with the first two commits (one code "suggestion" to appease check-typo). I agree with @gasche that it would be nice both to keep the diff quieter on the platform list and also be able to uncomment cases easily - possible additional commit to squash into the last one if you agree.

File: configure.ac
Comment (xavierleroy) [reply configure.ac]: Thanks for the hint!  (`m4_normalize`, indeed.)  Merged and squashed with the associated commit.

**Comment by xavierleroy** [comment]
I switched to `dnl` comments for the not-yet-supported configurations.  (Eventually I'll get comfortable with these comments.)

I didn't do anything special to have a non-commented case as the last case.  The generated configure script seems OK even with an extra empty default case at the end.

**Comment by xavierleroy** [comment]
Note to @shindere: the CI job I'm specifically trying to unbreak here is not the "main" job but the "other-configs" job, which currently fails at the first config it tries because `--disable-systhreads` is not correctly handled.

**Comment by xavierleroy** [comment]
Note to everyone: there may be other configurations that are not yet supported by 5.00 and should ideally not be configurable.  For example: bytecode-only on a 32-bit platform.

**Comment by dra27** [comment]
> Note to everyone: there may be other configurations that are not yet supported by 5.00 and should ideally not be configurable. For example: bytecode-only on a 32-bit platform.

That definitely wants to be fixed before release - for now I opened https://github.com/ocaml/ocaml/issues/10916 to track it in the 5.00.0 milestone.

**Review by dra27 — APPROVED** [review]
LGTM!

**Comment by xavierleroy** [comment]
Despite a bit of grumpiness about this PR, I'll take @dra27's approval and my own gut feeling that it is a (small) improvement as excuses to merge it now.

**Comment by xavierleroy** [comment]
And of course: stay tuned for more updates to the configure file as we figure out which features we can support in release 5.00 and which features need to be temporarily deactivated.

**Comment by shindere** [comment]
Xavier Leroy (2022/01/18 14:32 +0000):
> Note to @shindere: the CI job I'm specifically trying to unbreak here
> is not the "main" job but the "other-configs" job, which currently
> fails at the first config it tries because `--disable-systhreads` is
> not correctly handled.

Okay, sorry for having missed that point.

Shall we try to make the different tests independent of each other?

I guess this could be achieved either by creating one specific job per
configuration to be tried or, perhaps more lightly, by ignoring the
failures of the invocations of main in general.

**Pull request merged**
